cylc/uiserver/resolvers.py

```diff
@@ -321,6 +332,70 @@ async def play(cls, workflows, args, workflows_mgr, log):
         await workflows_mgr.scan()
         return response
 
+    @classmethod
+    async def enqueue(cls, stream, queue):
+        async for line in stream:
+            await queue.put(line.decode())
+
+    @classmethod
+    async def cat_log(cls, workflow: Tokens, log, info, task=None, file=None):
+        """Calls `cat log`.
+
+        Used for log subscriptions.
+        """
+        full_workflow = workflow
+        if file and task:
+            full_workflow = Tokens(f"{workflow.workflow_id}//{task}")
+        cmd = ['cylc', 'cat-log']
+        if file:
+            cmd += ['-f', file]
+        else:
+            cmd += ['-m', 't']
+        cmd.append(full_workflow.id)
+        log.info(f'$ {" ".join(cmd)}')
+
+        # For info, below subprocess is safe (uses shell=false by default)
+        proc = await asyncio.subprocess.create_subprocess_exec(
+            *cmd,
+            stdout=asyncio.subprocess.PIPE,
+        )
+        buffer: List[str] = []
+        queue: asyncio.Queue = asyncio.Queue()
+        # Farm out reading from stdout stream to a background task
+        # This is to get around problem where stream is not EOF until
+        # subprocess ends
+        asyncio.create_task(cls.enqueue(proc.stdout, queue))
+        op_id = info.root_value
+        if op_id not in info.context['sub_statuses']:
+            # subscription started in graphiql, add to dict
+            info.context['sub_statuses'][op_id] = 'start'
+        while True:
+            if (
+                op_id in info.context['sub_statuses']
+                and info.context['sub_statuses'][op_id] == 'stop'
+            ):
+                with suppress(psutil.Error):
+                    cat_log_proc = psutil.Process(proc.pid)
+                    # cat-log process will terminate on tail termination
+                    for tail_proc in cat_log_proc.children():
+                        tail_proc.terminate()
+                # clean the subscription from the status dict
+                info.context['sub_statuses'].pop(op_id)
+                break
```

Comment:
I think the `enqueue` coroutine will keep running unless explicitly cancelled, it won't shutdown automatically when it goes out of scope. It won't do anything much because it will be `await`ing lines from the `stream`, however, it will clog up the event loop as asyncio will still have to manage this coroutine (unless it bails with traceback because the `stream` isn't readable???).

We should explicitly stop it but cancelling async generators is a right pain with different behaviours at different Python versions annoyingly. I think `task.cancel()` [might do it](https://stackoverflow.com/a/60233813) e.g:

```python
enqueue_task = asyncio.create_task(cls.enqueue(proc.stdout, queue))
...
with suppress(...):  # suppress RuntimeError?
    enqueue_task.cancel()
```